File: backend/app/schemas/background_job.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import BackgroundJobStatus


class BackgroundJobResultSummaryResponse(BaseModel):
    """Summary metrics from the latest background job execution."""

    checked: int = Field(
        ge=0,
    )
    refreshed: int = Field(
        ge=0,
    )
    skipped: int = Field(
        ge=0,
    )
    failed: int = Field(
        ge=0,
    )

class BackgroundJobResponse(BaseModel):
    """Current state of a background job."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    key: str
    name: str
    schedule: str
    status: BackgroundJobStatus

    last_started_at: datetime | None
    last_finished_at: datetime | None
    last_duration_ms: int | None
    last_error: str | None
    next_run_at: datetime | None
    last_result: BackgroundJobResultSummaryResponse | None = None


class BackgroundJobRunResponse(BaseModel):
    """Execution history entry for a background job."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    job_id: UUID
    status: BackgroundJobStatus

    started_at: datetime
    finished_at: datetime | None
    duration_ms: int | None
    error: str | None
    result: dict[str, object] | None


class BackgroundJobRunNowResponse(BaseModel):
    """Accepted manual background job execution."""

    job: BackgroundJobResponse

File: frontend/lib/features/server/domain/models/background_job.dart
import 'package:equatable/equatable.dart';

enum BackgroundJobStatus { idle, running, success, failed }

final class BackgroundJobResultSummary extends Equatable {
  const BackgroundJobResultSummary({
    required this.checked,
    required this.refreshed,
    required this.skipped,
    required this.failed,
  });

  final int checked;
  final int refreshed;
  final int skipped;
  final int failed;

  @override
  List<Object?> get props => <Object?>[checked, refreshed, skipped, failed];
}

final class BackgroundJob extends Equatable {
  const BackgroundJob({
    required this.id,
    required this.key,
    required this.name,
    required this.schedule,
    required this.status,
    required this.lastStartedAt,
    required this.lastFinishedAt,
    required this.lastDurationMs,
    required this.lastError,
    required this.nextRunAt,
    required this.lastResult,
  });

  final String id;
  final String key;
  final String name;
  final String schedule;
  final BackgroundJobStatus status;

  final DateTime? lastStartedAt;
  final DateTime? lastFinishedAt;
  final int? lastDurationMs;
  final String? lastError;
  final DateTime? nextRunAt;

  final BackgroundJobResultSummary? lastResult;

  bool get isRunning {
    return status == BackgroundJobStatus.running;
  }

  bool get isHealthy {
    return status == BackgroundJobStatus.idle ||
        status == BackgroundJobStatus.success;
  }

  @override
  List<Object?> get props => <Object?>[
    id,
    key,
    name,
    schedule,
    status,
    lastStartedAt,
    lastFinishedAt,
    lastDurationMs,
    lastError,
    nextRunAt,
    lastResult,
  ];
}
